### PR TITLE
Allow users to customize dataloader

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #834
* __->__ #836

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -80,18 +78,15 @@ def main(job_config: JobConfig):
     )
     train_spec = get_train_spec(job_config.model.name)
 
-    # build tokenizer
-    tokenizer_type = model_name_to_tokenizer[train_spec.name]
-    tokenizer = build_tokenizer(tokenizer_type, job_config.model.tokenizer_path)
     # build dataloader
-    data_loader = build_hf_data_loader(
```
Comment (fduwjj) [inline train.py:87]: Nice!!

File: torchtitan/datasets/hf_datasets.py [hunk 2]
```diff
@@ -88,15 +88,15 @@ def __init__(
         ds = dataset_loader(path)
 
         self.dataset_name = dataset_name
-        self._data = split_dataset_by_node(ds, rank, world_size)
+        self._data = split_dataset_by_node(ds, dp_rank, dp_world_size)
```
Comment (fduwjj) [inline torchtitan/datasets/hf_datasets.py:90]: nice, this is indeed needed. When tracing the code.. I need to double check its definition in the caller, although the comment specified that.

File: torchtitan/dataloader.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    dp_rank: int
+    dp_world_size: int
+    batch_size: int
+
+    @abstractmethod
+    def __iter__(self):
+        ...
+
+
+class DPDataLoader(StatefulDataLoader, BaseDataLoader):
```
Comment (fduwjj) [inline torchtitan/dataloader.py]: I don't know but not really a big fan of this name. Maybe `ParallelAwareDataloader`?

**Review by fduwjj — APPROVED** [review]

File: torchtitan/dataloader.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    """Base class for all dataloaders.
+
+    This is used to enforce that all dataloaders have the methods defined in ``Stateful``,
+    ``state_dict()`` and ``load_state_dict()``.
+    """
+
+    tokenizer: Tokenizer
+    dp_rank: int
+    dp_world_size: int
+    batch_size: int
```
Comment (tianyu-l) [inline torchtitan/dataloader.py]: Not completely convincing to me if this is the right basic protocol:
1. What if people don't care if it supports checkpointing or not
2. what if people don't need it to be aware of DP ranks
3. Do all data loaders need to perform tokenization?

I feel that as along as it's an iterator, it's good enough. The things returned from the iterator don't need to be `input_ids, labels`. For sequence masking / multimodal, a dataloader needs to return more, e.g. mask, images, etc.

File: torchtitan/dataloader.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            "dataloader resharding is not supported yet."
+        )
+        # We don't have to use pickle as DCP will serialize the state_dict. However, we have to
+        # keep this for backward compatibility.
+        StatefulDataLoader.load_state_dict(
+            self, pickle.loads(state_dict[self._rank_id])
+        )
+
+
+class DataLoaderBuilder(Protocol):
```
Comment (tianyu-l) [inline torchtitan/dataloader.py]: ditto: is the signature of this protocol too strict?
Is it too relaxed if we substitute it to something like a `Callable` from anything to Iterator?

File: torchtitan/dataloader.py (hunk 4 above)
Comment (fegin) [reply torchtitan/dataloader.py]: I think item 2 and 3 are reasonable. And I actually think we can remove them. But what's the point of not supporting checkpointing?  For torchdata, which is a generic dataset library, it makes sense to have a very basic dataloader class. TorchTitan is a distributed training library, I don't see a reason why supporting checkpointing during training is not a must. 

Our checkpoint manager also assume dataloader to be Stateful. Removing Stateful is too relaxed, imo.

File: torchtitan/dataloader.py (hunk 5 above)
Comment (fegin) [reply torchtitan/dataloader.py]: For this one, an alternative in my mind was `Callable[[...], BaseDataloader]`.  This just relaxs the input and users can define whatever they want.

File: torchtitan/dataloader.py (hunk 4 above)
Comment (tianyu-l) [reply torchtitan/dataloader.py]: I think I'm more thinking from an experimentation platform perspective.
E.g. I'm a researcher working on a new type of attention. I have some data I'd like to load; I want to look at throughput gain; but I don't care about fault tolerance.

File: torchtitan/dataloader.py (hunk 5 above)
Comment (tianyu-l) [reply torchtitan/dataloader.py]: sounds good to me

File: torchtitan/dataloader.py (hunk 4 above)
Comment (fegin) [reply torchtitan/dataloader.py]: I'm not convinced by this use case. Users can simply raise `NotImplementedError` in `state_dict()` and `load_state_dict` of the dataloader. On the other hand, our checkpointer may still be used in this case to save the last model checkpoint. So even in such a research case, users may use the checkpointer which requires the dataloader to be `Stateful`. 

Maintaining the dataloader to be stateful will be more clean for most training loop. And for users who don't care about dataloader checkpointing, only extra ten lines of code are required.

Comment (tianyu-l) [reply torchtitan/dataloader.py]: > Users can simply raise NotImplementedError in state_dict() and load_state_dict of the dataloader. On the other hand, our checkpointer may still be used in this case to save the last model checkpoint. So even in such a research case, users may use the checkpointer which requires the dataloader to be Stateful.

Makes sense to me! Thanks!

File: torchtitan/dataloader.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    tokenizer: Tokenizer
+    dp_rank: int
+    dp_world_size: int
+    batch_size: int
+
+    def __init__(
+        self,
+        dataset: IterableDataset,
+        tokenizer: Tokenizer,
```
Comment (tianyu-l) [inline torchtitan/dataloader.py]: can we remove this, which is not used/needed

**Review by tianyu-l — APPROVED** [review]
LGTM -- please address the comment before merging
